Deprecate Geth miner namespace

e3243eric opened the pull request:
### What was wrong?

Closes #2856

Geth removed support for their [miner API](https://geth.ethereum.org/docs/interacting-with-geth/rpc/ns-miner).

### How was it fixed?

Change `GethMiner` methods to `DeprecatedMethod`.

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/master/newsfragments/README.md)

#### Cute Animal Picture

![Put a link to a cute animal picture inside the parenthesis-->](https://tourmeaway.com/blog/spots/Scouting_at_the_Taipei_Zoo_like_a_pro_Local_tip_to_the_ecology_of_Taiwan/cover2.jpg)

e3243eric committed:
Deprecate Geth miner namespace

reedsa commented on newsfragments/2857.removal.rst:
```diff
@@ -0,0 +1 @@
+Deprecate Geth miner namespace
```
Rename this file as `2857.deprecation.rst`

reedsa commented on tests/core/mining-module/test_miner_start.py:
```diff
@@ -26,7 +28,8 @@ def test_miner_start(w3_empty, wait_for_miner_start):
 
-    w3.miner.start(1)
+    with pytest.warns(DeprecationWarning):
+        w3.geth.miner.start(1)
```
Thanks for updating the namespace!

reedsa reviewed (commented):
One nit about the newsfragment filename, otherwise lgtm! @kclowes do you have any feedback here?

kclowes commented on web3/_utils/miner.py:
```diff
@@ -18,49 +19,62 @@
     mungers=[default_root_munger],
 )
 
+make_dag = DeprecatedMethod(_make_dag, "make_dag", "deprecated")
```
Nice use of `DeprecatedMethod`! I think since the only place we're using it is in this file, we can change it to better suit our purposes. Because we're not replacing the method with anything, we can update the warning [here](https://github.com/ethereum/web3.py/blob/main/web3/method.py#L250) to say something along the lines of f"{self.old_name} is no longer supported." 

Or, if you really wanted to go for some extra credit, you could change the `DeprecatedMethod` class to take a new, optional argument called `message` and then on `__get__` we could return the message if there is one. So the new DeprecatedMethod class might look something like:

```

class DeprecatedMethod:
    def __init__(
        self,
        method: Method[Callable[..., Any]],
        old_name: str,
        new_name: Optional[str] = None,
        msg: Optional[str] = None,
    ) -> None:
        self.method = method
        self.old_name = old_name
        self.new_name = new_name
        self.msg = msg

    def __get__(
        self, obj: Optional["Module"] = None, obj_type: Optional[Type["Module"]] = None
    ) -> Any:
        if self.msg is not None:
            warnings.warn(self.msg)
        else:
            warnings.warn(
                f"{self.old_name} is deprecated in favor of {self.new_name}",
                category=DeprecationWarning,
            )
        return self.method.__get__(obj, obj_type)
```

And this line would look something like:
```suggestion
make_dag = DeprecatedMethod(_make_dag, "make_dag", msg="All mining methods have been deprecated")
```

kclowes reviewed (commented):
Thanks for reviving this PR, @reedsa! 

@e3243eric I made a comment around the messaging in the `DeprecatedMethod` class, let me know what you think. I'm happy to make that update if you don't want/have time to, just let me know!

Thanks for taking it on!

fselmo committed:
Initial commit for middleware refactor

- Refactor middlewares as classes with request processors and response processors
- Apply this refactor to some existing middlewares and test
  - gas_price_strategy_middleware
  - validation / formatting middleware
  - attrdict_middleware
  - (incomplete) ens_name_to_address_middleware
- Refactor the exception_retry_middleware to be a configuration on the http providers

fselmo committed:
Refactor new middleware onion setup

fselmo committed:
Refactor caching as a decorator on ``make_request()``

- Remove all caching middleware.
- Refactor request caching via configurations on the provider classes and decorators on their respective ``make_request()`` methods.

fselmo committed:
Remove middleware dependent on old logic

- This middleware depends on a request always being made when middleware are processed. This isn't how the middleware is processed for websockets and the separation of middleware into request param processing and response processing further helps with being able to support batch requests. Remove all middleware that were dependent on a request always being sandwiched between request processing and response processing.

fselmo committed:
Add request mocker fixture for mocking requests

- Add a ``RequestMocker`` class with ``request_mocker`` pytest fixture to fascilitate request mocking now that the middleware doesn't sequester the ``make_request()`` call. Since we can't swap the ``make_request()`` on the fly, mock results and errors using this test utility class.
- Linting
- Some cleanup along the way

fselmo committed:
Some refactoring to make cached_requests work with WSV2

fselmo committed:
WIP: Fix or comment out core test errors so core suite can run

fselmo committed:
Allow for manipulating the method via middleware

- In the middleware refactor, we weren't accounting for manipulating the ``method``, only the params. Allow for this kind of manipulation as it is necessary for ``sign_and_send_raw_middleware``, for example.

fselmo committed:
Re-introduce the sandwiched middleware model via a method on the middleware class

- Use a repr() for the middleware onion name when a non-hashable, such as the curried `build` method on a web3middlewarebuilder is used.

fselmo committed:
Build on the wrapped make_request middleware refactor

- Add support back for ``local_filter_middleware`` by overriding the ``_wrap_make_request()`` of the base middleware class. If this method is overridden, the ``make_request`` can be effectively replaced if a request is not needed.

fselmo committed:
Fix remaining tests that were broken due to refactor

fselmo committed:
Remove the necessity for the ``abi_middleware``

- closes #2995
- ``ens_name_to_address_middleware`` would return the ``bytes`` address, rather than checksummed address. Since this happens after the ``request_formatters`` have been applied, this would override the results from the ``ABI_REQUEST_FORMATTERS`` which take ``address`` types and format them to a checksummed hex address.

fselmo committed:
Reinstate the http retry request tests as configuration

fselmo committed:
Fix type hinting for refactor

fselmo committed:
Fix inconsistency with eth-tester integration test

fselmo committed:
Remove unnecessary imports hidden by # noqa: F401

fselmo committed:
use request mocker + no need to format logsBloom in eth-tester middleware

- Use request_mocker where we can
- result_processors take care of the formatting... this was causing the ``None`` value for logsBloom to raise when mocked

fselmo committed:
Rename all references of geth_poa to extradata_to_poa

- The middleware changed names in the refactor, but this commit puts the finishing touches and closes #899

fselmo committed:
Minor cleanups from refactor PR

fselmo committed:
Update eth-tester and add support for eth_feeHistory

fselmo committed:
'name_to_address' -> 'ens_name_to_address' for consistency.

fselmo committed:
Make caching more robust; add tests back for cached requests

- Make request caching more robust by re-introducing the lock from the middleware as well as checking for "error" and None responses and not caching those.
- Re-introduce the tests for the simple_cache_middleware as caching utils tests.

[Unrelated]

- Keep the decorator logic on ``make_request`` when using the ``request_mocker`` fixture so we can still utilized the request mocker for testing cached requests.

fselmo committed:
Use a message listener background task over ws.recv()

- Get rid of ``websockets.recv()`` calls in favor of a background task
  that listens for messages and puts them into appropriate queues /
  caches.

- This rids us of the need to have a ws_lock() around the ``recv()``
  call.

- Each provider has a listener task that error logs unless the user
  specifies an exception limit. In the case one is specified, exceptions
  will be raised and hault the listener task.

- In the (ideally) rare case that the subscription queue fills up to its
  max size, this will trigger an asycio.Event that will cause the
  listener task to wait until messages are consumed from the queue to
  resume listening.

fselmo committed:
Make listener task exception handling a boolean

fselmo committed:
Improve INFO logging and naming for new wsv2 setup

- Use INFO logging for communicating the state of the request processor
  caching.

- If the subscription message queue is in sync with the  websocket
  provider, refrain from INFO logging each message and instead log
  that we are in sync with the websocket message stream. If we go out
  of sync, log that we are out of sync and log the number of messages
  up to the point of sync.

- Rename some methods and classes to better reflect their purpose now
  that we no longer pull directly from the websocket but use a listener
  task and internally managed queue.

fselmo committed:
Add tests around message listener task; some refactor + renaming

fselmo committed:
``listen_to_websocket`` -> ``process_subscriptions`` renaming:

- Since we always have a listener background task that listens to the websocket now,
we should be more honest in the naming for this method. What it does is listen to
subscription messages that may or may not be synced with the websocket message stream.
Some of these may already be in the cache.

fselmo committed:
Update documentation around the websocket provider v2 refactor

fselmo committed:
``pass`` instead of INFO logging for expected exceptions.

fselmo committed:
Only yield active subscriptions.

fselmo committed:
Pass ``PersistentConnectionProvider`` kwargs through

fselmo committed:
Add a test for listen Event when subs queue is full

fselmo committed:
Add newsfragment for #3179

fselmo committed:
Keep the message listener alive inside a while True

fselmo committed:
Changes from comments on PR #3169

Nov1kov committed:
Fix types - AsyncContractConstructor

fselmo committed:
Validate geth version in test fixture against binary used for tests

fselmo committed:
newsfragment for #3193

fselmo committed:
Use as_tuple_of_middlewares() to get around typing issues

- Use an internal method for the ``NamedMiddlewareOnion`` to generate
  the tuple of middlewares, instead of relying on ``tuple()``.

fselmo committed:
newsfragment for 3169

fselmo committed:
Drop support for python 3.7:

We will have more peripheral updates to make due to the dropping of
python 3.7, such as forgoing some typing imports from typing_extensions,
etc. For now, we have PRs that are already expecting python 3.7 to be
dropped and all tests fail there. This commit aims to only drop the tests,
the CI support for py37, and the claim that the library supports 3.7.2 and above.

fselmo committed:
newsfragment for #3198

fselmo committed:
closes #2884 - drop lru-dict dependency

fselmo committed:
newsfragment for #3196

fselmo committed:
Update documentation related to the middleware refactor:

- Add documentation for the new middleware model.
- Remove async versions of middleware from the docs; this is no longer
  a necessary distinction.
- Remove the `normalize_request_parameters` middleware from the docs and
  codebase. This wasn't used anywhere internally.

fselmo committed:
address comments from #3195

fselmo committed:
Changes from comments on #3179

fselmo committed:
Raise exceptions by default; opt to silence in message listener task

fselmo committed:
newsfragment for #3202

fselmo committed:
re-compile test contracts with solidity ``v0.8.24``

fselmo committed:
newsfragment for #3204

fselmo committed:
Update documentation for ``WebsocketProviderV2`` + minor refactor.

- Remove the ``endpoint_uri`` argument from the ``PersistentConenctionProvider`` base class, opting for defining it in the inheriting classes instead.
- Update the documentation around the ``WebsocketProviderV2`` class to reflect the current state + document ``PersistentConenctionProvider`` under a new section.

Commit:
Formatting updates for black v24.1.0

pacrob committed:
restore class order and add newsfragment

Commit:
newsfragment for #3207

fselmo committed:
Update setup for HTTP rexception retry config

fselmo committed:
Update documentation on caching and exception retries for v7

fselmo committed:
Start the ``v6`` -> ``v7`` migration guide

fselmo committed:
Relevant newsfragments for #3211

fselmo committed:
Clarify how custom middleware can be added to the stack

- Clarify that middleware can be added to the onion either via the class
  or, in the case of the default middleware, a reference to the class.
- Minor typo fixes for docs updates.

fselmo committed:
Remove references to middleware classes; opt for class-only

fselmo committed:
Fix issues with missing repr of unhashables in middleware onion

fselmo committed:
Final cleanups after removing middleware references to classes

fselmo committed:
``geth --dev`` test fixture setup (#3191)

* WIP: Generate geth fixture for shanghai with clique consensus

* Initial squashed commits for getting ``geth --dev`` test fixture working:

- Quicker timeout for unmined wait-for-txn-receipt tests
- Update default fees test for geth --dev mode
- Get rid of all miner start() and stop() references
- Run integration tests with --dev flag
- Test refactoring for ``geth --dev`` test fixture setup
- Remove PoW related endpoint tests; TODO: remove / deprecate endpoints
- Support latest geth versions
- Update geth version in circleCI config + peripherally related changes
- Turn on PoS block identifier tests
- get geth --dev test suite working

* Updates towards ``geth --dev`` test fixture:

- Flaky for replace transaction
- Fix typing
- Loosen assertions on gas
- Nonce value fix
- Loosen asserts for default maxFeePerGas and maxPriorityFeePerGas tests

* ``geth --dev`` test fixture tweaks:

- Add @flaky to tests that expect mining at certain times
- This isn't ideal. Perhaps we can increase the ``dev.period`` (mining
  interval) to make these tests a bit more reliable and hopefully
  the other tests are unaffected.
- Update benchmark to run with ``geth --dev`` setup
- Put back old state of get_logs_without_logs for eth_tester

* More work towards ``geth --dev`` test fixture:

- Try with --dev.period=5
- Remove assert from test_eth_new_block_filter
- Remove sleep from test_web3_client_version
- Try without flaky
- Fixes test_eth_send_transaction_with_nonce without the need for retries.
- Retries were causing transactions to be sent with the same nonce,
  which in turn requires more gas to go through. Running the test once
  fixes this issue.
- Obtain the transaction count providing the 'pending' argument.
- Fix test_eth_send_transaction_no_max_fee

* Additional cleanup for ``geth --dev`` test fixture

* Address feedback for ``geth --dev`` test fixture

---------

Co-authored-by: Stuart Reed <stuart.reed@ethereum.org>

fselmo committed:
changes from comments on PR #3211

fselmo committed:
Catch newly raised geth errors when tx indexing in progress:

- Geth ``1.13.11`` started to return an error when transaction indexing
is in progress and a ``eth_getTransactionReceipt`` call is made.
Handle this in the ``wait_for_transaction_receipt`` method by catching
the error in order to continue waiting.

fselmo committed:
newsfragment for #3216

fselmo committed:
Minor update to docstrings for #3216

Commit:
Return function in contract function iterator (#3200)

* Contract function iterator returns function itself

* Core tests for function iterator

* Update news

fselmo committed:
newsfragment for #3172

fselmo committed:
format as code

Commit:
Estimate Gas with State Override (#3168)

* Estimate gas with state override param

* `state_override` docs and test improvements

Update test to pass None for block identifier so geth ignores it, per note in web3.contract.rst

Update description for `state_override`` param in docs

Add async test for `state_override` in estimate gas

Combine logic for eth_call and eth_estimateGas mungers

Mark not implemented in eth tester

* Update newsfragment for #3168

fselmo committed:
Add formatters for Cancun block header fields:

- Add formatters for "blobGasUsed", "excessBlobGas", and "parentBeaconBlockRoot"
- Add fields to block api tests

fselmo committed:
newsfragment for #3223

fselmo committed:
``CallOverride`` -> ``StateOverride``

- State overrid is used in more requests than just ``eth_call``.
  We should use a more generic name that suits this type more.
  ``StateOverride`` seems the most appropriate here.

fselmo committed:
news for #3227

fselmo committed:
Add note for CallOverride -> StateOverride to v7 migration

kclowes committed:
Move newsfragment to newsfragment directory (#3229)

Commit:
Contract Events Sorted by `ContractEvent.logIndex` and `ContractEvent.blockNumber` (#3228)

* Sort events by logIndex

* Test for contract event logs to ensure sort order

* Add doc for get_logs sort order

* Update sort by block number and log index

* newsfragment for #3228

kclowes committed:
Upgrade geth fixture (#3231)

* Rename fixture file, updates to circle config

* Add newsfragment

ymittal committed:
chore: fix typos

Commit:
`DeprecatedMethod` class takes `msg` kwarg for custom warnings

Commit:
Newsfragment for #2857

pacrob reviewed (approved):
lgtm! - but should be to v6, removed for v7, yes?

fselmo commented:
> lgtm! - but should be to v6, removed for v7, yes?

Yep, along with #3236, this should close it all out

reedsa commented:
Rebased this against `main` by mistake, closing in favor of #3238
